Source/WebCore/platform/audio/cocoa/PlatformRawAudioDataCocoa.cpp

```diff
@@ -0,0 +1,320 @@
+{
+    return isAudioSampleFormatInterleaved(format) ? CAAudioStreamDescription::IsInterleaved::Yes : CAAudioStreamDescription::IsInterleaved::No;
+}
+
+static RetainPtr<CMSampleBufferRef> createSampleBuffer(const CAAudioStreamDescription& description, const CMTime& time, size_t numberOfFrames, WebAudioBufferList* list)
+{
+    CMAudioFormatDescriptionRef rawFormatDescription;
+    if (PAL::CMAudioFormatDescriptionCreate(kCFAllocatorDefault, &description.streamDescription(), 0, nullptr, 0, nullptr, nullptr, &rawFormatDescription))
+        return nullptr;
+    RetainPtr formatDescription = adoptCF(rawFormatDescription);
+
+    CMSampleBufferRef rawSampleBuffer;
+    if (PAL::CMAudioSampleBufferCreateWithPacketDescriptions(kCFAllocatorDefault, nullptr, false, nullptr, nullptr, rawFormatDescription, numberOfFrames, time, nullptr, &rawSampleBuffer))
+        return nullptr;
+    auto sampleBuffer = adoptCF(rawSampleBuffer);
+
+    if (list) {
+        if (PAL::CMSampleBufferSetDataBufferFromAudioBufferList(sampleBuffer.get(), kCFAllocatorDefault, kCFAllocatorDefault, kCMSampleBufferFlag_AudioBufferList_Assure16ByteAlignment, list->list()))
+            return nullptr;
+        return sampleBuffer;
+    }
+
+    auto result = WebAudioBufferList::createWebAudioBufferListWithBlockBuffer(description, numberOfFrames);
+    if (!result)
+        return nullptr;
+
+    auto [newList, blockBuffer] = WTFMove(*result);
+    if (PAL::CMSampleBufferSetDataBuffer(rawSampleBuffer, blockBuffer.get()))
+        return nullptr;
+    return sampleBuffer;
+}
+
+RefPtr<PlatformRawAudioData> PlatformRawAudioData::create(std::span<const uint8_t> sourceData, AudioSampleFormat format, float sampleRate, int64_t timestamp, size_t numberOfFrames, size_t numberOfChannels)
+{
+    CAAudioStreamDescription inputDescription(sampleRate, numberOfChannels, audioSampleFormatToPCMFormat(format), interleavedFormat(format));
+    if (!WebAudioBufferList::isSupportedDescription(inputDescription, numberOfFrames)) {
+        RELEASE_LOG_ERROR(MediaStream, "PlatformRawAudioData::create not supported configuration");
+        return nullptr;
+    }
+
+    WebAudioBufferList inputList = inputDescription;
+    WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+    uint8_t* data = const_cast<uint8_t*>(sourceData.data());
+    size_t sizePlane = numberOfFrames * inputDescription.bytesPerFrame();
+    for (auto& buffer : inputList.buffers()) {
+        buffer.mData = data;
+        buffer.mNumberChannels = inputDescription.numberOfInterleavedChannels();
+        buffer.mDataByteSize = sizePlane;
+        data += sizePlane;
+    }
+    if (data > sourceData.data() + sourceData.size()) {
+        RELEASE_LOG_ERROR(MediaStream, "PlatformRawAudioData::create nonsensical format data");
+        return nullptr;
+    }
+    WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+
+    RetainPtr sample = createSampleBuffer(inputDescription, PAL::CMTimeMake(timestamp, 1000000), numberOfFrames, &inputList);
```

Comment:
Why passing inputList as a pointer?

Reply:
hmm, in a previous iteration of the code the list was optional, but it now longer is. So I'll make it a reference. good spotting